### PR TITLE
Add reindex option to write atom id to gro file

### LINKED ISSUE #1777: Changing atom id and write out
### Expected behaviour
I have a file which I wish to change the atom id according to some rules. I wonder what is the correct way of altering this property? Thanks.

```python
from MDAnalysis import Universe
from six import StringIO
gro = '''test
1
    2CL      CL20850   0.000   0.000   0.000
7.29748 7.66094 9.82962'''
u = Universe(StringIO(gro), format='gro')
u.atoms[0].id = 3
u.atoms.write('temp.gro')
#temp.gro
#test
#1
#    2CL      CL3       0.000   0.000   0.000
#7.29748 7.66094 9.82962
```
### Actual behaviour
The atom id is not changed to 3.
And the output gro file gives
```python
#temp.gro
#test
#1
#    2CL      CL1       0.000   0.000   0.000
#7.29748 7.66094 9.82962
```

### Currently version of MDAnalysis:
py36/mda0.17.0

### PR DESCRIPTION
**Pull request by xiki-tempula**
Fixes #1777 
Can I have some advice on which doc should I change? 
Should I only change https://www.mdanalysis.org/docs/documentation_pages/coordinates/GRO.html
or should I mention it on other writer pages? 
Suggestions are welcomed.

Changes made in this Pull Request:
 - 
```python
from MDAnalysis import Universe
from six import StringIO
gro = '''test
1
    2CL      CL20850   0.000   0.000   0.000
7.29748 7.66094 9.82962'''
u = Universe(StringIO(gro), format='gro')
u.atoms[0].id = 3
u.atoms.write('temp.gro', reindex=False)
```
Gives
```
test
1
    2CL      CL    3   0.000   0.000   0.000
7.29748 7.66094 9.82962
```

While
```python
from MDAnalysis import Universe
from six import StringIO
gro = '''test
1
    2CL      CL20850   0.000   0.000   0.000
7.29748 7.66094 9.82962'''
u = Universe(StringIO(gro), format='gro')
u.atoms[0].id = 3
u.atoms.write('temp.gro', reindex=True)
```
Gives
```
test
1
    2CL      CL    1   0.000   0.000   0.000
7.29748 7.66094 9.82962
```

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

File: testsuite/MDAnalysisTests/coordinates/test_gro.py [hunk 1]
```diff
@@ -418,6 +419,30 @@ def test_growriter_resid_truncation():
     # larger digits should get truncated
     assert line.startswith('56789UNK')
 
+@tempdir.run_in_tempdir()
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/coordinates/test_gro.py]: this should be two tests

File: package/MDAnalysis/coordinates/GRO.py [hunk 2]
```diff
@@ -377,7 +387,7 @@ def write(self, obj):
             # all attributes could be infinite cycles!
             for atom_index, resid, resname, name in zip(
                     range(ag_or_ts.n_atoms), resids, resnames, names):
-                truncated_atom_index = util.ltruncate_int(atom_index + 1, 5)
+                truncated_atom_index = util.ltruncate_int(atom_indexs[atom_index], 5)
```
Comment (richardjgowers) [inline package/MDAnalysis/coordinates/GRO.py]: I'd put `atom_indexs` into the `zip` call to make it similar to `resid` etc

File: package/MDAnalysis/coordinates/GRO.py [hunk 3]
```diff
@@ -344,6 +345,15 @@ def write(self, obj):
         except (AttributeError, NoDataError):
             resids = itertools.cycle((1,))
             missing_topology.append('resids')
+
+        if not self.reindex:
+            try:
+                atom_indexs = ag_or_ts.ids
```
Comment (richardjgowers) [inline package/MDAnalysis/coordinates/GRO.py]: the plural of index is indices, it's a weird exception

File: package/MDAnalysis/coordinates/GRO.py (hunk 2 above)
Comment (xiki-tempula) [reply package/MDAnalysis/coordinates/GRO.py]: @richardjgowers atom_index is used Later in the code as an index
pos=positions[atom_index]
vel=velocities[atom_index]

Thus, if I zip the atom_indexs in to the zip, I will need to create another index start from 0.

File: testsuite/MDAnalysisTests/coordinates/test_gro.py [hunk 4]
```diff
@@ -418,6 +419,57 @@ def test_growriter_resid_truncation():
     # larger digits should get truncated
     assert line.startswith('56789UNK')
 
+class TestGrowriterReindex():
+    def __init__(self):
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/coordinates/test_gro.py]: We don't use init methods for tests.  If you want to reuse `self.u` make it into a pytest fixture

File: testsuite/MDAnalysisTests/coordinates/test_gro.py [hunk 5]
```diff
@@ -418,6 +419,57 @@ def test_growriter_resid_truncation():
     # larger digits should get truncated
     assert line.startswith('56789UNK')
 
+class TestGrowriterReindex():
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/coordinates/test_gro.py]: should subclass from object for Py2/3 compatibility

File: testsuite/MDAnalysisTests/coordinates/test_gro.py [hunk 6]
```diff
@@ -418,6 +419,57 @@ def test_growriter_resid_truncation():
...
+
+    @tempdir.run_in_tempdir()
+    def test_growriter_resid_true(self):
+        self.u.atoms.write('temp.gro', reindex=True)
+
+        with open('temp.gro', 'r') as grofile:
+            grofile.readline()
+            grofile.readline()
+            line = grofile.readline()
+        # larger digits should get truncated
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/coordinates/test_gro.py]: this comment (and repeats below) aren't useful

**Comment by richardjgowers** [comment]
@xiki-tempula I've fixed up the tests you've added, pytest is a little weird sometimes.  If you've got any questions about the changes ping me

**Comment by xiki-tempula** [comment]
@richardjgowers Thank you very much. I have struggled a bit to use this pytest.fixture, where the test seems to fail in weird places. I guess I have now learnt how to use it.

**Comment by richardjgowers** [comment]
@xiki-tempula thanks for the fix!

**Pull request merged**